### PR TITLE
Adding equivalent to numpy.nan_to_num functionality

### PR DESCRIPTION
**Pull request by aspeers**
Adding equivalent to numpy.nan_to_num functionality
(https://numpy.org/doc/stable/reference/generated/numpy.nan_to_num.html)
addressing issue 479 (https://github.com/pymc-devs/pytensor/issues/479).

### Motivation for these changes
Requested in issue #479

### Implementation details
Added nan_to_num in pytensor/tensor/math.py

### Checklist
+ [x] Explain motivation and implementation 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues, preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes). Note that if they don't, we will [rewrite/rebase/squash the git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) before merging.
+ [ ] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## New features
- added nan_to_num

## Bugfixes
- ...

## Documentation
Example:

```
import pytensor
from pytensor import tensor as pt

import numpy as np

# Replace NaN
print("Replace NaN")
x = pt.dvector("x")
y = pt.log(x)

f = pytensor.function([x], y)

t = np.random.normal(size=(10,))
print("x")
print(t)
print("f = log(x)")
print(f (t))

# Test: NaN default replacement
print("\nTest: NaN default replacement")
y1 = pt.math.nan_to_num(y)
f = pytensor.function([x], y1)
print(f (t))

# Test: NaN custom replacement
print("\nTest: NaN custom replacement")
y2 = pt.math.nan_to_num(y, nan=1)
f = pytensor.function([x], y2)
print(f (t))

# Replace +INF/-INF
print("\n\nReplace +INF/-INF")
a = pt.dvector("a")
b = pt.dvector("b")
c = pt.math.true_div(a,b)

t_a = np.random.normal(size=(10,))
t_b = np.zeros((10,))

f = pytensor.function([a, b], c)
print("a")
print(t_a)
print("b")
print(t_b)
print("f = a/b")
print(f(t_a, t_b))

# Test: +INF/-INF default replacement
c1 = pt.math.nan_to_num(c)
f = pytensor.function([a, b], c1)
print("\nTest: +INF/-INF default replacement")
print(f(t_a, t_b))

# Test: +INF/-INF custom replacement
c2 = pt.math.nan_to_num(c, posinf=5, neginf=-5)
f = pytensor.function([a, b], c2)
print("\nTest: +INF/-INF custom replacement")
print(f(t_a, t_b))
```

## Maintenance
- ...

### CODE REVIEW

**Comment by aspeers** [comment]
First PR from yesterday's code sprint.  Main reasons for marking as in progress included:

- Had some difficulty running pytest locally.  Resulted in multiple failures (most involving missing files/config parameters) which I figured the community could help by providing the relevant pytest command to run.
- Could use advice on adding unit tests for the addition.
- Was suggested to use ifelse() by @jessegrabowski instead of traditional if statements.  ifelse is currently not imported/used in the math.py file and adding it to the imports list was causing a circular reference.  Any thoughts would be welcomed!

File: pytensor/tensor/math.py [hunk 1]
```diff
@@ -2937,7 +2937,58 @@ def matmul(x1: "ArrayLike", x2: "ArrayLike", dtype: Optional["DTypeLike"] = None
...
+    # Specify the value to replace +INF and -INF with
+    if posinf is not None:
+        maxf = posinf
+    if neginf is not None:
+        minf = neginf
+
+    # Replace +INF and -INF values
+    pos = gt(sign(x), 0)
+    x = switch(bitwise_and(isinf(x), pos), maxf, x)
+    x = switch(bitwise_and(isinf(x), ~pos), minf, x)
```
Comment (ricardoV94) [inline pytensor/tensor/math.py:2985]: Let's compare directly with posinf and neginf, if we don't have a `isposinf` and `isneginf` it's a good time to add it.

```suggestion
    x = switch(bitwise_and(isinf(x), pos), maxf, x)
    x = switch(bitwise_and(isinf(x), ~pos), minf, x)
```

File: pytensor/tensor/math.py [hunk 2]
```diff
@@ -2937,7 +2937,58 @@ def matmul(x1: "ArrayLike", x2: "ArrayLike", dtype: Optional["DTypeLike"] = None
...
+        The tensor with NaN's, +INF, and -INF replaced with the
+        specified and/or default substitutions.
+    """
+
+    # Replace NaN's with nan keyword
+    x = switch(isnan(x), nan, x)
+
+    # Get max and min values representable by x.dtype
+    maxf = np.finfo(x.real.dtype).max
+    minf = np.finfo(x.real.dtype).min
```
Comment (ricardoV94) [inline pytensor/tensor/math.py:2974]: Let's define it only when posinf and neginf are None

File: pytensor/tensor/math.py [hunk 3]
```diff
@@ -2937,7 +2937,58 @@ def matmul(x1: "ArrayLike", x2: "ArrayLike", dtype: Optional["DTypeLike"] = None
     return out
 
 
+def nan_to_num(x, nan=0.0, posinf=None, neginf=None):
+    """
+    Replace NaN values with the `nan` keyword, +INF with the `posinf`
```
Comment (ricardoV94) [inline pytensor/tensor/math.py:2942]: Let's just copy Numpy docstring. No point in reinventing it

File: pytensor/tensor/math.py [hunk 4]
```diff
@@ -2937,7 +2937,58 @@ def matmul(x1: "ArrayLike", x2: "ArrayLike", dtype: Optional["DTypeLike"] = None
...
+
+    Returns
+    -------
+    out
+        The tensor with NaN's, +INF, and -INF replaced with the
+        specified and/or default substitutions.
+    """
+
+    # Replace NaN's with nan keyword
+    x = switch(isnan(x), nan, x)
```
Comment (ricardoV94) [inline pytensor/tensor/math.py:2970]: Instead of overriding `x`, let's write all the checks `isnan`\`isposinf`\`isneginf` on the original array.

**Review by ricardoV94 — COMMENTED** [review]
Thanks for opening the PR. Looks great!

I left some small technical suggestions. We will also need some tests

**Comment by ricardoV94** [comment]
`ifelse` is not appropriate because it requires a scalar condition, and here we want to work with arbitrary tensors

**Comment by Dhruvanshu-Joshi** [comment]
Hi @aspeers , are you still working on this?

**Pull request closed without merging**